python/models/hierarchical_filtering_proxy_model.py

```diff
@@ -267,56 +277,50 @@ def filterAcceptsRow(self, src_row, src_parent_idx):
+        traversed to reject a particular item.
+
+        A rejection cache is build when this method is called on top nodes to
+        avoid to compute the same values again and again when this method is called
+        in nodes deeper in the hierarchy. This cache keeps track of pruned branches
+        in the tree.
 
         :param src_row:         The row in the source model to filter
         :param src_parent_idx:  The parent index in the source model to filter
         :returns:               True if the row should be accepted by the filter, False
                                 otherwise
         """
-        # get the source index for the row:
+        # Get the source index for the row:
         src_model = self.sourceModel()
         src_idx = src_model.index(src_row, 0, src_parent_idx)
 
-        # first, see if any children of this item are known to already be accepted
-        child_accepted = self._child_accepted_cache.get(src_idx)
-        if child_accepted == True:
-            # child is accepted so this item must also be accepted
-            return True
-
-        # next, we need to determine if the parent item has been accepted.  To do this,
-        # search up the hierarchy stopping at the first parent that we know for sure if
-        # it has been accepted or not.
-        upstream_indexes = []
-        current_idx = src_idx
-        parent_accepted = False
-        while current_idx and current_idx.isValid():
-            accepted = self._accepted_cache.get(current_idx)
-            if accepted != None:
-                parent_accepted = accepted
-                break
-            upstream_indexes.append(current_idx)
-            current_idx = current_idx.parent()
-
-        # now update the accepted status for items that we don't know
-        # for sure, working from top to bottom in the hierarchy ending
-        # on the index we are checking for:
-        for idx in reversed(upstream_indexes):
-            accepted = self._is_row_accepted(idx.row(), idx.parent(), parent_accepted)
-            self._accepted_cache.add(idx, accepted)
-            parent_accepted = accepted
-
-        if parent_accepted:
-            # the index we are testing was accepted!
-            return True
-        elif src_model.hasChildren(src_idx):
-            # even though the parent wasn't accepted, it may still be needed if one or more
-            # children/grandchildren/etc. are accepted:
-            return self._is_child_accepted_r(src_idx, parent_accepted)
+        # Check if we are dealing with top nodes
+        if not src_parent_idx.parent() or not src_parent_idx.parent().isValid():
+            self._update_rejection_cache(src_idx)
+            # We can now just return the cached value which was may be modified
+            # when traversing the tree.
+            cached_value = self._rejected_cache.get(src_idx)
+            return not cached_value
```

Comment:
Minor: the `not cached_value` is confusing especially considering the three return values `True` `False` or `None`.

Could we do something like

```
_IndexRejectedCache.is_accpted(index) # raises KeyError or something in cache if value not present?
_IndexRejectedCache.is_rejected(index) # raises KeyError or something in cache if value not present?
_IndexRejectedCache.is_cached(index)
```

Perhaps that leads to clearer code?